### PR TITLE
Add AssetActive model

### PR DESCRIPTION
**Pull request by uranusjr**
This tracks asset orphaning like the old is_orphaned flag (but in the other direction, as the name suggests). The difference is that this table also serves as a unique constraint on both the name and URI used by the user is unique across a DAG deployment. This is different from the unique constraint on Asset, which only ensures the name-URI combination is unique.

I also took the chance to also change AssetAlias’s `name` length limit to 1500 to match Asset’s. Not necessary at all, but this probably removes some possibility to confusion? I can remove it if people feel otherwise.

### CODE REVIEW

**Comment by jscheffl** [comment]
I don't understandwhy we need another table for this function. Seems to be adding too much complexity. Why not just adding the unique index and a flag for active (potentially with an index on active flag)?

**Comment by uranusjr** [comment]
The addition of `name` to Asset makes the unique constraint more complex because we want to keep history of different name-URI combinations. For example, if you have two assets (in the history of the Airflow deployment) at the same S3 location, but used for different purposes, we want to keep the two distinct for history purposes. Conversely, if an asset name is linked to different URIs, we want to keep thw entries distinct too. This means the unique constraint on Asset should be against the `(name, uri)` combination.

However, during one DAG-processing round, we also want to keep one name only used once—otherwise it would be extremely confusing if the user accidentally uses the same name to represent different assets in different DAGs. So we also need to have a unique constraint _only on name if an asset is active_. This is possible with Postgres with `create unique index ... on asset (name) WHERE (is_orphaned = false)`, but not anywhere else. This is what the new table is for—when an asset is in use, it has an entry in AssetActive, and that table enforces uniqueness on `name` (and `uri`, same problem) individually. This essentially also replaces `is_orphaned` (an orphaned asset will not have a corresponding entry in AssetActive), so `is_orphaned` is removed.

**Comment by jscheffl** [comment]
> The addition of `name` to Asset makes the unique constraint more complex because we want to keep history of different name-URI combinations. For example, if you have two assets (in the history of the Airflow deployment) at the same S3 location, but used for different purposes, we want to keep the two distinct for history purposes. Conversely, if an asset name is linked to different URIs, we want to keep thw entries distinct too. This means the unique constraint on Asset should be against the `(name, uri)` combination.
> 
> However, during one DAG-processing round, we also want to keep one name only used once—otherwise it would be extremely confusing if the user accidentally uses the same name to represent different assets in different DAGs. So we also need to have a unique constraint _only on name if an asset is active_. This is possible with Postgres with `create unique index ... on asset (name) WHERE (is_orphaned = false)`, but not anywhere else. This is what the new table is for—when an asset is in use, it has an entry in AssetActive, and that table enforces uniqueness on `name` (and `uri`, same problem) individually. This essentially also replaces `is_orphaned` (an orphaned asset will not have a corresponding entry in AssetActive), so `is_orphaned` is removed.

Mhm, okay. But adding the table complexity just because of DB limitations... still I'd feel a handling on app level is "cheaper" than adding an additional shadow table just to push down the constraint handling to DB. That also adds a lot of potential inconsistency and redundancy on DB level.

Not a fan but if others are OK with I don't want to block this. I just "hurts my eyes" as a former DBA to see the model.

Other questions: Is the requirement of the history needed like this or could also the history implemented in another way w/o the need of non-uniqueness?

**Comment by uranusjr** [comment]
The problem is we can’t do this at the app side. DAG processing may be done in multiple processes, in which cases there’s no single process that has access to know if a name or URI is unique, and the database is the last line to keep the input sane. I guess we can choose to not do the extra table, and always just use `session.scalar(...)` to get _one_ (whichever) object and hope things are somewhat consistent, but I’m not quite sure if that’d be simpler, to be honest.

**Comment by uranusjr** [comment]
Oh hey, I managed to fix everything!

File: airflow/models/asset.py [hunk 1]
```diff
@@ -232,6 +232,61 @@ def to_public(self) -> Asset:
...
+            refcolumns=["dataset.name", "dataset.uri"],
+            name="asset_active_asset_name_uri_fkey",
+            ondelete="CASCADE",
+        ),
+        Index("idx_asset_active_name_unique", name, unique=True),
+        Index("idx_asset_active_uri_unique", uri, unique=True),
+    )
+
+    @classmethod
+    def for_asset(cls, asset: AssetModel) -> AssetActive:
```
Comment (Lee-W) [inline airflow/models/asset.py:286]: What does `for_asset` mean?

File: airflow/jobs/scheduler_job_runner.py [hunk 2]
```diff
@@ -2066,8 +2066,13 @@ def _orphan_unreferenced_assets(self, session: Session = NEW_SESSION) -> None:
             )
         )
 
-        updated_count = sum(self._set_orphaned(asset) for asset in orphaned_asset_query)
-        Stats.gauge("asset.orphaned", updated_count)
+        orphaning_identifiers = list(self._get_orphaning_identifier(asset) for asset in orphaned_asset_query)
```
Comment (Lee-W) [inline airflow/jobs/scheduler_job_runner.py]: why do we want to use `list` instead of `[]` here?

File: airflow/models/asset.py (hunk 1 above)
Comment (Lee-W) [reply airflow/models/asset.py:286]: from_asset?

```suggestion
    def from_asset(cls, asset: AssetModel) -> AssetActive:
```

File: airflow/jobs/scheduler_job_runner.py (hunk 2 above)
Comment (uranusjr) [reply airflow/jobs/scheduler_job_runner.py]: I think I simply forgot to change it’s a thing (it’s easier to change `sum` to `list`…)

I’ll use a list comprehension instead.

File: airflow/models/asset.py (hunk 1 above)
Comment (uranusjr) [reply airflow/models/asset.py:286]: I used `for` because this creates a unique constraint _for_ the asset…? I’m not strong on either.

Comment (Lee-W) [reply airflow/models/asset.py:286]: oh... no strong opinion either

**Review by Lee-W — APPROVED** [review]
One nitpick not addressed, but I think we're good even if that's not changed

**Comment by uranusjr** [comment]
I made the list comprehension anyway (and rebased).

**Pull request merged**